A ~17x speed optimization of creating tensor from multiple numpy arrays using torch.tensor and torch.Tensor.

touqir14 opened the pull request:
**Fixes** [13918](https://github.com/pytorch/pytorch/issues/13918) , [39842](https://github.com/pytorch/pytorch/issues/39842).
Creating a tensor using ```torch.tensor``` from a list (or lists of lists and so on) of numpy arrays is quite slow compared to just doing ```torch.tensor(numpy.stack(list_of_arrays))```. This PR optimizes this very particular case of having (only) numpy arrays stacked recursively arbitrarily deep inside a list/tuple. It can be [x1, x2,....] or even [[[x1, x2], [x3, x4]], [[x5, x6], [x7, x8]]] , where x_i are numpy arrays. This contribution not only improves the speed but also makes ```torch.tensor``` and ```tensor.Torch``` at least as fast (usually faster) than doing ```torch.tensor(numpy.stack(list_of_arrays))```.

The following example illustrates the performance improvements:
```
import torch

import numpy as np

a1 = np.random.randn(10000)
list_a1 = [a1 for _ in range(10000)]

%timeit -n 15 -r 10 torch.tensor(list_a1) # With Optimization
313 ms ± 328 µs per loop (mean ± std. dev. of 10 runs, 15 loops each)

%timeit -n 15 -r 10 torch.tensor(np.stack(list_a1))
1.11 s ± 26.3 ms per loop (mean ± std. dev. of 10 runs, 15 loops each)

%timeit torch.tensor(list_a1) # Without Optimization
5.31 s ± 100 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)



a2 = np.random.randn(100,100)
list_a2 = [a2 for _ in range(10000)]

%timeit -n 15 -r 10 torch.Tensor(list_a2) # With Optimization
308 ms ± 530 µs per loop (mean ± std. dev. of 10 runs, 15 loops each)

%timeit -n 15 -r 10 torch.tensor(np.stack(list_a2))
1.65 s ± 120 ms per loop (mean ± std. dev. of 10 runs, 15 loops each)

%timeit torch.tensor(list_a2) # Without Optimization
5.79 s ± 93.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

```
In the above case, it can be seen that the optimized ```torch.tensor``` is at least 3.5x faster than using the Numpy stacking array approach and at least ~17x faster than the previous version of ```torch.tensor```.
This is my first PR to PyTorch's repository and I am still wrapping my head around the code base. Please do let me know of any shortcomings. I will be adding test cases to the [test_tensor_creation_ops.py](https://github.com/pytorch/pytorch/blob/master/test/test_tensor_creation_ops.py) file (this is the right place is it?) and do let me know if I should write tests anywhere else.

#### TODO:
- [ ] Write test cases.

touqir14 committed:
Optimizing tensor creation from multiple numpy arrays - updated tensor_new.cpp

touqir14 committed:
Optimizing tensor creation from multiple numpy arrays - updated tensor_numpy.cpp

touqir14 committed:
Added tests to test_tensor_creation_ops.py

touqir14 commented:
Added basic tests. Let me know if more is needed.

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,54 @@ def test_inference(default_dtype):
+        """
+
+        repeats = 10
+        np_types = [np.int8, np.int16, np.int32, np.int64, np.float16, np.float32,\
```
Instead of creating its own NumPy type list this test can reuse the keys in this dict:

https://github.com/pytorch/pytorch/blob/929b91a24d7c78483cf9c0dc434810e640d06a2c/torch/testing/_internal/common_utils.py#L382

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,54 @@ def test_inference(default_dtype):
+
+        for dtype in np_types: # Simple test
+            for _ in range(repeats):
+                arrays = [(10*np.random.randn(100)).astype(dtype) for _ in range(repeats)]
```
Why multiply by 10?

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,54 @@ def test_inference(default_dtype):
+        for dtype in np_types: # Simple test
+            for _ in range(repeats):
+                arrays = [(10*np.random.randn(100)).astype(dtype) for _ in range(repeats)]
+                self.assertEqual(torch.tensor(arrays, device=device), torch.tensor(np.stack(arrays), device=device), exact_dtype=True, rtol=1e-05, atol=1e-08)
```
Are there more distinct code paths to validate the correctness of torch.tensor() when given ndarrays? Maybe a Python reference implementation that converts each NumPy array to a tensor first? Or using torch.from_numpy()?

touqir14 replied:
Its a little cumbersome. A different path than using numpy.stack would have been to use torch's equivalent function but its stack function does not support recursively adding numpy arrays within lists. I changed ```torch.tensor``` to ```torch.from_numpy```

mruberry commented:
Hey @touqir14, thanks for submitting this PR. Looks like there are some compilation issues preventing the CI from running properly. Would you take a look at the test failures?

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,54 @@ def test_inference(default_dtype):
+                self.assertEqual(torch.tensor(arrays, device=device), torch.tensor(np.stack(arrays), device=device), exact_dtype=True, rtol=1e-05, atol=1e-08)
+
+
+        for dtype in np_types: # Simple test
```
This test is very repetitious. Just create the arrays up front and up them in an iterable, then iterate over the product of them and the NumPy dtypes, cast the arrays to the NumPy dtype, and perform the test by comparing torch.tensor() to a reference (like from_numpy() or a custom reference).

Make sure there are empty arrays, and 0D-3D arrays in the test set, too. It's also great that you're looking at arrays which are not contiguous or views.

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -279,10 +293,21 @@ Tensor internal_new_from_data(
     return tensor.to(device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/copy_numpy);
   }
+
+  std::vector<PyArrayObject*> array_ptr_storage;
```
Add a comment explaining this block

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -279,10 +293,21 @@ Tensor internal_new_from_data(
   }
+
+  std::vector<PyArrayObject*> array_ptr_storage;
+  if (torch::utils::extract_ndarrays(data, sizes, array_ptr_storage)) {
```
Refactor this block so it creates the tensor and acquires the scalar type, but then code common to both this block and the next block acquires the device, release the gil, possibly initializes CUDA, calls to() on the tensor and returns it. (This will require a small edit to the next block, too.)

touqir14 replied:
Its done.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
 }
+
+
+template <typename T>
```
Add a comment explaining this function and how it's intended to be used

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  if (end < 0) {
+    end = vec.size();
+  }
+  for (int i=start; i<end; ++i) {
```
match i's type to start's type

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    end = vec.size();
+  }
+  for (int i=start; i<end; ++i) {
+    output += (std::to_string(vec[i]) + ", ");
```
Won't this put an extra comma at the end of the tuple?

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+
+template <typename T>
+std::string get_vector_str(std::vector<T> & vec, int64_t start, int64_t end=-1){
+  std::string output = "(";
```
use an ostream to construct text output

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  return output + ")";
+}  
+
+template <typename T>
```
Add a comment explaining this function and how it's intended to be used

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+}  
+
+
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, size_t (&numels)[], size_t shape_idx, std::vector<PyArrayObject*> & array_ptr_storage) {
```
Add a comment explaining this function and how it's intended to be used

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  size_t numels[ndims];
+  numels[ndims-1] = shape[ndims-1];
+
+  for (int i=ndims-2; i>=0; --i) {
```
match i's type to ndim's type
put spaces around the assignment operator and the geq

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+}
+
+
+bool extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<PyArrayObject*> & array_ptr_storage) {
```
Add a comment explaining this function and how it's intended to be used

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    if (numels[shape_idx] != PyArray_SIZE((PyArrayObject*) obj)) {
+      auto true_shape_str = get_buffer_str<npy_intp> (static_cast<void*>(PyArray_DIMS((PyArrayObject*) obj)), 0, PyArray_NDIM((PyArrayObject*) obj)) + std::string(")");
+      auto expected_shape_str = get_vector_str(shape, shape_idx);
+      auto err_str = std::string("expected numpy array of shape ") + expected_shape_str + std::string(" at dim ") + std::to_string(shape_idx) + std::string("(got ") + true_shape_str;
```
What does an example error message look like?

touqir14 replied:
This : ```ValueError: expected numpy array of shape (20, 30) at dim 1(got (30, 40))```

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+      auto true_shape_str = get_buffer_str<npy_intp> (static_cast<void*>(PyArray_DIMS((PyArrayObject*) obj)), 0, PyArray_NDIM((PyArrayObject*) obj)) + std::string(")");
+      auto expected_shape_str = get_vector_str(shape, shape_idx);
+      auto err_str = std::string("expected numpy array of shape ") + expected_shape_str + std::string(" at dim ") + std::to_string(shape_idx) + std::string("(got ") + true_shape_str;
+      throw ValueError(err_str);  
```
Use TORCH_CHECK

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, size_t (&numels)[], size_t shape_idx, std::vector<PyArrayObject*> & array_ptr_storage) {
+  if (PyArray_Check(obj)) {
+    if (numels[shape_idx] != PyArray_SIZE((PyArrayObject*) obj)) {
```
Add a comment explaining what's happening in this block. The error logic really dominates.

Is it sufficient to just check the number of elements? Isn't the shape of the array relevant?

touqir14 replied:
I actually gave a bit of a leeway. As long as the number of elements are the same as the expected number of elements the results will be correct since the raw memory copy routines sees a numpy array and its flattened version in the same way.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  } 
+  else if (PySequence_Check(obj)) {
+    auto seq = THPObjectPtr(PySequence_Fast(obj, "not a sequence"));
+    if (!seq) throw python_error();
```
Egyptian-style curly braces for every if/for statement so the linter doesn't complain

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    array_ptr_storage.push_back((PyArrayObject*) obj);
+    return true;
+  } 
+  else if (PySequence_Check(obj)) {
```
same line as closing curly brace

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    auto seq = THPObjectPtr(PySequence_Fast(obj, "not a sequence"));
+    if (!seq) throw python_error();
+    auto seq_len = PySequence_Fast_GET_SIZE(seq.get());
+    if (seq_len != shape[shape_idx]) {
```
use TORCH_CHECK

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+        (long long)shape[shape_idx], (long long)shape_idx, (long long)seq_len);
+    }
+    PyObject** items = PySequence_Fast_ITEMS(seq.get());
+    for (size_t i=0; i<seq_len; ++i) {
```
match i's type to seq_len's type

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+
+
+template <typename T>
+void memcpy_cpu(void* dst, const void* src, size_t num_items, bool isaligned, int itemsize=1) {
```
Add a comment explaining this function and how it's intended to be used

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+        char* dst_tmp = ((char*) dst) + begin;
+        auto last = src_tmp+(end-begin);
+
+        while(src_tmp < last){
```
How would this handle discontiguous arrays? Like an array with an innermost stride that isn't the bitwidth of its dtype? (Put another way, what about a view of an array with every other element missing?)

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+        T* dst_tmp = ((T*) dst) + begin;
+        auto last = src_tmp+(end-begin);
+
+        while(src_tmp < last){
```
Same question as below re: discontiguous arrays / arrays with "gaps"

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+}
+
+
+Tensor reverse_strides(Tensor & tensor) {
```
Comment here

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+}
+
+
+at::Tensor tensor_from_ndarray_batch(std::vector<PyArrayObject*> & array_ptr_storage, std::vector<int64_t> & sizes, ScalarType scalarType, bool pin_memory) {
```
Comment here

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  auto size = strides.size();
+  std::vector<long int> new_strides(size);
+  new_strides[0] = strides[0];
+  for (int i=0; i<size-1; ++i) {
```
type match the loop var

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+Tensor reverse_strides(Tensor & tensor) {
+  auto strides = tensor.strides();
+  auto size = strides.size();
+  std::vector<long int> new_strides(size);
```
don't use long int, either type match another type or use an exact type (like int64_t)

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+}
+
+
+inline void* store_ndarray(void* tensor_ptr, void* ndarray_ptr, at::ScalarType scalarType, size_t num_items, int itemsize, bool isaligned) {
```
comment here

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+      C_contiguous_count += 1;
+  }
+
+  if (2*C_contiguous_count < array_ptr_storage.size())
```
curly braces here and on other if and for blocks, even if they only contain a single statement

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  int np_type = aten_to_numpy_dtype(scalarType);
+  auto tensor_ptr = tensor.data_ptr();
+
+  double time = 0;
```
this is probably a debugging leftover?

mruberry reviewed (commented):
I left some inline questions and comments for your review, @touqir14. Looking forward to hearing your thoughts.

touqir14 committed:
Added comments and slight modifications to code and tests

touqir14 committed:
few changes to pass lint test

touqir14 committed:
few changes to pass lint test

touqir14 committed:
should pass flake tests

touqir14 commented:
@mruberry , I have made almost all of the suggested changes. Regarding the TORCH_CHECK, is there any advantage to using it instead of raising ValueError? I thought the latter would be more informative since the errors are really about the input.
Regarding the case of contiguous arrays, the call to ```PyArray_FromArray``` ensures that we are dealing with contiguous and aligned arrays in case they are not. The overhead for non-contiguous arrays is not much and I don't think it is worthwhile to add the logic to cover all cases for a small performance gain. Its better to leave the internal implementation hassles to the numpy API.

touqir14 commented:
I am not sure why ```clang-tidy``` is giving this error : ```/home/runner/work/pytorch/pytorch/torch/csrc/utils/tensor_new.cpp:50:50: error: use of undeclared identifier 'PyArrayObject' [clang-diagnostic-error]``` . The proper numpy header files are included, so why would this show up?

touqir14 commented:
Not sure why this error is showing :  ```RuntimeError: test_cpp_api_parity failed! Received signal: SIGSEGV``` from https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.0.1-py3.6-test1/131/console

touqir14 committed:
fixed bugs

touqir14 committed:
Enabled default type setting for the tensor_from_ndarray_batch approach

touqir14 commented:
@mruberry , did you get a chance to look into the updates?

ngimel commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +355,260 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  This utility function is used for copying elements from a raw numpy array memory to a tensor memory location by using memcpy_cpu and proper type matching.
+  */
+  switch (scalarType) {
+    case at::kByte: memcpy_cpu<uint8_t> (tensor_ptr, ndarray_ptr, num_items, itemsize, isaligned); return (void*)(((uint8_t *)tensor_ptr) + num_items);
```
you should be able to use dispatch macro here instead of spelling out all the dtypes. Bonus points for using OpaqueType (see in ATen/native/cuda/IndexKernel.cu) to minimize number of template instantiations, but the benefit may be too small to bother.

touqir14 replied:
Could you please refer me to where I can find this macro, or any source file where it has been used? Still getting used to the codebase.

ngimel replied:
For an example use see e.g. here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/IndexKernel.cu#L210. Dispatch macro defines C type `scalar_t` depending on scalarType, that you can use to instantiate templated functions, it is implemented pretty much like this switch statement that you have here.

ngimel commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +355,260 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+          Py_DECREF((PyObject*) array_tmp);
+        }
+    } else {
+        auto array_tmp = (PyArrayObject*) PyArray_FromArray(array, PyArray_DescrFromType(np_type), NPY_ARRAY_FARRAY);
```
all discontiguous cases (not just f-contiguous) go here, which leads to some unexpected strides in the result:
```
In [16]: a=np.random.rand(2,3,4)[:,:,0]                                                                                                                                                                      

In [17]: a.strides                                                                                                                                                                                           
Out[17]: (96, 32)

In [18]: lists=[a for _ in range(10)]                                                                                                                                                                        

In [19]: b=torch.tensor(lists)                                                                                                                                                                               

In [20]: b.stride()                                                                                                                                                                                          
Out[20]: (6, 1, 3)
```

touqir14 replied:
"all discontiguous cases (not just f-contiguous) go here" : That part was intentional. There was a flaw in the way the strides were assigned whenever any non C-contiguous array were used. It somehow eluded the test cases that I wrote. I have changed the way the strides are computed whenever C-contiguous arrays are in minority. For the majority case, I don't have to change the strides since any non C-contiguous arrays are mapped into a C-contiguous array. I have also added extra tests cases including the one that you just wrote above. All tests are passing. Let me know if I should add any other tests.

touqir14 replied:
I have limited the optimization to the case that the input list contains numpy arrays of the same shape being the leaves at the same level. The same shape and same leaf level constraints are required for this new stride assigning strategy to work correctly.

ngimel replied:
I'm unsure this `else` branch makes sense. It avoids array copy in `PyArray_FromArray` if the input lists consists of (mostly) f-contiguous arrays, which is likely a rarely encountered niche case. All other discontiguous inputs will first be copied into f-contiguous form, and then sent to store_ndarray. The resulting pytorch tensor will be discontiguous, which, again, likely is not intended outcome. If this `else` branch didn't exist, then those discontiguous cases would be copied to c-contiguous, and result would be contiguous tensor, which is same amount of memory traffic and arguably better output. So, roughly, now:
1) c-contiguous inputs -> no copy by PyArray_FromArray, contiguous output (good!)
2) f-contiguous inputs -> no copy by PyArray_FromArray in the `else` branch, discontiguous output (good, but niche)
3) discontiguous inputs -> copy by PyArray_FromArray in the else branch, discontiguous output (unexpected). 
If you remove `else` branch
1) c-contiguous inputs -> no copy by PyArray_FromArray, contiguous output (good!)
2) f-contiguous inputs -> copy by PyArray_FromArray, contiguous output (worse than above, but code is also simpler)
3) discontiguous inputs -> copy by PyArray_FromArray, contiguous output (same memory traffic as above, better output format). 
Your thoughts?
Also, can you add test cases for discontiguous inputs if they are not present already?

touqir14 replied:
Let me state the logic here and the rationale behind it. The idea is that the input can consist of c-contiguous, f-contiguous and discontiguous arrays and that we need to either map all these arrays into aligned, c-contiguous / f-contiguous arrays since some of the arrays might not be aligned (considering all cases). The majority voting strategy has been used for obvious reasons to minimize the extra conversion overhead. Two reasons for converting to c-contiguous/f-contiguous arrays : 

1. The ```store_ndarray``` copying mechanism assumes that the arrays are contiguous (c or f).
2. The resulting numpy arrays must all have the same memory layout - either all being c-contiguous or all being f-contiguous. This is required due to way the strides of the tensor are computed. 

Currently, with the newest commit, here is what happens:
1. mostly c-contiguous inputs -> no copy by PyArray_FromArray, c-contiguous output (As you said earlier). Any non c-contiguous input will be made c-contiguous (by copying if necessary).
2. mostly f-contiguous inputs -> no copy by PyArray_FromArray in the else branch, f-contiguous output. Any non f-contiguous (that includes c-contiguous and discontiguous arrays) input will be made f-contiguous (by copying if necessary). While f-contiguous inputs are less common, the most common case is the transpose of a c-contiguous numpy array. 
3. With the latest commit : discontiguous inputs -> copy by PyArray_FromArray, c-contiguous output.

Now if I remove the ```else``` branch, any f-contiguous array needs to be mapped to a c-contiguous format even if all of them are f-contiguous . All of them might be the transpose of a c-contiguous array. Now that I have gotten the striding strategy right, I am a bit inclined to keep the ```else``` branch.

Test cases for c-contiguous, f-contiguous and discontiguous arrays are present currently.

touqir14 committed:
Used a different strategy for assigning strides for F-contiguous arrays

touqir14 committed:
specialized the optimization to arrays of exact same shape

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,88 @@ def test_inference(default_dtype):
+        for dtype in np_types:
+            for _ in range(repeats):
+                stack = []
+                A = np.random.randn(100).astype(dtype)
```
Make sure this also tests simple conversions, like a single empty tensor, and tensors with gaps in the memory, like converting tensors where every other element is missing

touqir14 replied:
I am not sure what you meant by "converting tensors where every other element is missing". Could you explain further? Like having strides that jump over the odd entries?

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,88 @@ def test_inference(default_dtype):
+        for dtype in np_types:
+            for _ in range(repeats):
+                stack = []
+                A = np.asfortranarray(np.random.randn(100).astype(dtype))
```
What about a case that mixes f and c contiguous (and discontiguous) tensors?

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -126,10 +134,17 @@ Tensor new_with_tensor(c10::DispatchKey dispatch_key, at::ScalarType scalar_type
   std::vector<int64_t> sizes;
   THPObjectPtr handle;
   while (PySequence_Check(seq)) {
+    if (PyArray_Check(seq) && PyArray_NDIM((PyArrayObject*) seq) == 0) { 
```
Let's also add an explicit test for [a] and [a, b] and [a, c] where a and b are scalar arrays and c is not.

touqir14 replied:
Having a mixture of scalar and non scalar arrays in the list will throw an exception - this logic was always there (in the ```recursive_store```). However, having only scalar arrays is perfectly fine.

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -259,6 +277,11 @@ Tensor internal_new_from_data(
     return var.to(device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/copy_variables);
   }
 
+  bool use_batched_array_extraction = false;
```
Why is this block declared here? It's just doing extra work if the data was an array.

touqir14 replied:
I pulled out a couple of the parts that were common to https://github.com/pytorch/pytorch/blob/82f0e3ca45d4e52fe5ef782c7c8d318700e328ab/torch/csrc/utils/tensor_new.cpp#L311-L314 and https://github.com/pytorch/pytorch/blob/82f0e3ca45d4e52fe5ef782c7c8d318700e328ab/torch/csrc/utils/tensor_new.cpp#L321-L328 as you asked earlier. If you want I can put the lines ```auto sizes = compute_sizes(data);``` and ```ScalarType inferred_scalar_type = type_inference ? infer_scalar_type(data) : scalar_type;``` back into these two blocks to avoid the extra computation when the data is an array.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+bool extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  A function for extracting all numpy arrays embedded (recursively) within PySequences in obj. Failure to do so will return
+  false, else true. The extracted numpy array pointers are stored in array_ptr_storage. It uses _extract_ndarrays() function 
```
Why would this return false?

touqir14 replied:
The only time it will return true is when all the items are non-scalar numpy arrays having the same shapes and also all of them need to be the leaf nodes at the same height.  So, ```[[a,b], c]``` will return false as ```a``` and ```c``` are at different heights.

mruberry replied:
Thank you, that's helpful. Please add that to the comment, too.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t shape_idx, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  This is a helper function for extracting all the numpy arrays embedded within PySequences (recursively) representd by obj. This is done recursively in a 
+  depth first search style. It puts the numpy array pointers in the array_ptr_storage for later use. In case there is a non numpy array type within obj, 
```
Does it only return false if there are non-NumPy objects?

touqir14 replied:
yeah. See https://github.com/pytorch/pytorch/pull/51731#discussion_r582556575

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  */
+  if (PyArray_Check(obj)) {
+    if (PyArray_NDIM((PyArrayObject*) obj) == 0) {
+      return false;
```
This seems to violate the contract stipulated in the comment that it returns true if all the objects are NumPy arrays?

touqir14 replied:
I must have forgotten to update the comment.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  return true;
+}
+
+void _populate_shape_vector(std::vector<int64_t> & shape, PyArrayObject* array) {
```
This function is only called in one place so just inline it

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  */
+  int64_t ndim = PyArray_NDIM(array);
+  npy_intp* array_shape = PyArray_DIMS(array);
+  for (int64_t i = 0; i < ndim; ++i) {
```
This should be using a standard library call to copy

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+      ndarray_shape.emplace_back(shape_idx);
+      _populate_shape_vector(ndarray_shape, (PyArrayObject*) obj);
+    }
+    else if ((ndarray_shape[0] != shape_idx) || !_shape_match_ndarray(ndarray_shape, (PyArrayObject*) obj, 1)) {
```
This conditional is hard to parse -- what's actually being tested for here?

touqir14 replied:
The first part evaluates to be true if there exists an array embedded within the sequence at a different height than the current array pointed by ```obj```. The second part evaluates to be true if there exists an array with a different shape than the array pointed by ```obj```.

mruberry commented:
Hey @touqir14, @ngimel and I took a look for a bit but we have to leave off reviewing for the moment.

The code path this PR is editing is very fragile, so we want to be very careful about changes, and we think this PR is doing a good job of making a surgical change there.

We still have some concerns about test coverage, in particular covering corner cases and validating contiguity, but those should be straightforward to address.

We're also struggling a bit with all the new functions. They are tricky to read. Maybe there's a way to simplify them? What about, for example, focusing exclusively on C contiguous arrays in a single list for now, for example? Maybe that would let us focus more on the testing and integration, and then we could extend fast path availability in the future if we wanted.

touqir14 commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,88 @@ def test_inference(default_dtype):
+        A = np.random.rand(2, 3, 4)[:, 0, :]
+        stack.append([A for _ in range(10)])
+
+        for arrays in stack:
```
Perhaps we can add a keyword argument for controlling whether the output must be C-contiguous and set its default value to be true? Currently, it is guaranteed to produce a C-contiguous tensor only if the majority of the input numpy arrays are not F-contiguous. A trivial modification can enforce a behaviour that under all cases, C-contiguous tensors will be generated at the cost of a runtime overhead whenever that keyword argument is true.

mruberry replied:
Let's not add a kwarg (because changing the signature is a little risky) and just preserve the current behavior for now. We can revisit changing behavior or extending functionality in the future. This PR is focused on improving performance.

touqir14 commented:
I think the main issue is with ```_extract_ndarrays``` and reducing its complexity should not be hard. A major chunk of that function is devoted to checking the properties of the numpy arrays and deciding whether to fall back to the slower path. If we delegate that responsibility to a helper function and let ```_extract_ndarrays``` to just recursively traverse the list and add the numpy arrays to the ```array_ptr_storage``` based on the decisions of that helper function, that will be helpful. In terms of complexity, ```_extract_ndarrays``` would be similar to ```recursive_store``` : https://github.com/pytorch/pytorch/blob/e2b4f55e658a236cfabe827b67850cf388eeb088/torch/csrc/utils/tensor_new.cpp#L220-L242
From there, I think we can step by step cover all possible test cases. I think it would be great to be able to cover the optimizations for C-contiguous, f-contiguous and discontiguous numpy arrays in this PR rather than leaving a part of it for later submission. @mruberry , @ngimel

mruberry commented:
> I think the main issue is with `_extract_ndarrays` and reducing its complexity should not be hard. 

That'd be helpful. 

> From there, I think we can step by step cover all possible test cases. I think it would be great to be able to cover the optimizations for C-contiguous, f-contiguous and discontiguous numpy arrays in this PR rather than leaving a part of it for later submission.

OK, but let's be sure we're solving a real problem and not a hypothetical one where we're given a motley collection of NumPy arrays.

touqir14 commented:
@mruberry , I think the following cases represent real world uses: 

1. A list of C-contiguous arrays.
2. The extension of (1) for at least a few layers deep like [[[x1,x2],[x3,x4]],[[x5,x6],[x7,x8]]] and alike cases. I think such use cases typically arise for real time prediction tasks like video classification using 3D CNNs where the time window can vary and the image frames are produced as numpy arrays realtime.
3. Use of transposed arrays along with C-contiguous arrays in case (1) and (2). One case could be that the images are recorded from multiple cameras having different point-of-views and for prediction purposes the images need to be transposed, rotated, flipped, etc. Thus we are dealing with some number of f-contiguous arrays here.
4. Mixture of (f and/or c) contiguous and discontiguous arrays for case (1) and (2) which are common if the frames are coming from multiple cameras with different resolutions.

The current state of this PR includes optimizations for all of the above cases.

mruberry commented:
> @mruberry , I think the following cases represent real world uses:
> 
> 1. A list of C-contiguous arrays.
> 2. The extension of (1) for at least a few layers deep like [[[x1,x2],[x3,x4]],[[x5,x6],[x7,x8]]] and alike cases. I think such use cases typically arise for real time prediction tasks like video classification using 3D CNNs where the time window can vary and the image frames are produced as numpy arrays realtime.
> 3. Use of transposed arrays along with C-contiguous arrays in case (1) and (2). One case could be that the images are recorded from multiple cameras having different point-of-views and for prediction purposes the images need to be transposed, rotated, flipped, etc. Thus we are dealing with some number of f-contiguous arrays here.
> 4. Mixture of (f and/or c) contiguous and discontiguous arrays for case (1) and (2) which are common if the frames are coming from multiple cameras with different resolutions.
> 
> The current state of this PR includes optimizations for all of the above cases given the constraint that the shapes of the arrays must be the same and they all must be at the same level (not going to work for [[x1, x2], x3] and alike cases) and that all of the leaf lists will contain the same number of arrays (not going to work for [[x1,x2],[x3]] and alike cases)

OK, I can be convinced. Let us know when this is ready for another review.

touqir14 commented:
I should be able to send commits to address all the remaining suggestions and changes before the end of this week. I will let you know.

touqir14 committed:
refactored and added few more tests

touqir14 committed:
fixed flake error

touqir14 committed:
Merge remote-tracking branch 'upstream/master'

touqir14 commented:
@mruberry , @ngimel , I have updated this PR with new commits. I have refactored ```_extract_ndarrays``` in order to decrease its complexity. ```store_ndarray``` now relies upon one of the dispatching macros which has lead to conciseness. Few more test cases has been added one of which tests c-contiguous, f-contiguous and discontiguous inputs together.

touqir14 commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -279,15 +302,23 @@ Tensor internal_new_from_data(
+  // In case data contains anything other than all numpy arrays, this approach will not be taken and
+  // instead recusrive_store as shown below will be used for copying.
+  std::vector<PyArrayObject*> array_ptr_storage;
+  if (torch::utils::extract_ndarrays(data, sizes, array_ptr_storage)) {
```
Done

touqir14 commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -259,6 +277,11 @@ Tensor internal_new_from_data(
     return var.to(device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/copy_variables);
   }
 
+  bool use_batched_array_extraction = false;
```
Done

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -42,6 +46,10 @@ using at::Type;
 
 namespace torch { namespace utils {
+
+at::Tensor tensor_from_ndarray_batch(std::vector<PyArrayObject*> & array_ptr_storage, std::vector<int64_t> & sizes, ScalarType scalarType, bool pin_memory);
```
Instead of declaring these here put them in a header.

touqir14 replied:
Ah, I actually forgot to move those into a header.

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -284,15 +307,27 @@ Tensor internal_new_from_data(
+  // We first check whether the numpy array pointers were extracted. If so, we continue with raw memory
+  // copying part : torch::utils::tensor_from_ndarray_batch.
+  // In case data contains anything other than all numpy arrays, this approach will not be taken and
+  // instead recusrive_store as shown below will be used for copying.
```
recusrive_store -> recursive_store

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -264,6 +283,10 @@ Tensor internal_new_from_data(
     return var.to(device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/copy_variables);
   }
 
+  auto sizes = compute_sizes(data);
```
Add a comment here about why sizes and inferred_scalar_type is computed here.

Ideally they would be computed closer to where they're used, since the code immediately below these definitions shadows infered_scalar_type and doesn't use size.

touqir14 replied:
I have moved the computation of ```infered_scalar_type``` into the blocks where it is being used.

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -284,15 +307,27 @@ Tensor internal_new_from_data(
+    at::tracer::impl::NoTracerDispatchMode tracer_guard;
+    tensor = torch::utils::tensor_from_ndarray_batch(array_ptr_storage, sizes, inferred_scalar_type, pin_memory);
+    if (as_contiguous) {
+      tensor = tensor.contiguous();
```
Would matching the behavior today require calling contiguous()?

touqir14 replied:
As per https://github.com/pytorch/pytorch/pull/51731#discussion_r593522402 , I have removed this part.

mruberry commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -716,10 +752,10 @@ void _validate_sparse_coo_tensor_args(c10::DispatchKey dispatch_key, at::ScalarT
 Tensor tensor_ctor(c10::DispatchKey dispatch_key, at::ScalarType scalar_type, PyObject* args, PyObject* kwargs) {
   static PythonArgParser parser({
-    "tensor(PyObject* data, *, ScalarType dtype=None, Device? device=None, bool pin_memory=False, bool requires_grad=False, DimnameList? names=None)",
+    "tensor(PyObject* data, *, ScalarType dtype=None, Device? device=None, bool pin_memory=False, bool requires_grad=False, DimnameList? names=None, bool as_contiguous=True)",
```
Let's not modify this function signature.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+
+  if (ndarray_shape.size() == 0) {
+    /*
+    If ndarray_shape is an empty vector, then ndarray_shape[0] is set to the depth of the array in the iterable and
```
This reuse of one data structure for two things is a little confusing for the reader. 

I'm a little confused why the algorithm doesn't work conceptually like this:

- recurse down leftmost branches of the forest until it finds an array, identifies the (common) depth and (common) shape of that array, and then returns those values (either by modifying inputs or returning a tuple)
- then discover each array in the forest and validate it has the properties of the first array

The current algorithm may be actually be doing something similar to this, but it's tricky to read.

touqir14 replied:
```_validate_ndarray``` first finds out the depth and shape of the first numpy array it sees and then in each subsequent calls it validates whether the rest of the arrays have same shape and depth. So yes the current algorithm does that using a single function. Doing this using a single function is more concise than breaking it apart. I have replaced the single datastructure with a pair where pair.first is the depth and pair.second is the vector that represents the target shape.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  return output.str();
+}  
+
+inline bool _shape_match_ndarray(std::vector<int64_t> & shape, PyArrayObject* array, int64_t start) {
```
This function looks fine but it might be simplified by removing the "start" value if the data structures used are refined as suggested below.

touqir14 replied:
Removed start.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    npy_intp* array_shape = PyArray_DIMS(array);
+    ndarray_shape.insert(ndarray_shape.end(), array_shape, array_shape + ndim);
+  }
+  else if ((ndarray_shape[0] != depth) || !_shape_match_ndarray(ndarray_shape, array, 1)) {
```
Use 

https://github.com/pytorch/pytorch/blob/5b648ef909fbf89c53f28ebc1b3bd2f4fde168c5/c10/util/Exception.h#L435

touqir14 replied:
Replaced it.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    }
+    Py_ssize_t seq_len = PySequence_Fast_GET_SIZE(seq.get());
+    if (seq_len != shape[depth]) {
+      throw ValueError("expected sequence of length %lld at dim %lld (got %lld)",
```
Same comment here as above about using the existing macros vs. directly throwing errors

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  of a dimension and/or depth mismatch (See _validate_ndarray for more info). This function is used in extract_ndarrays() function below.
+  */
+
+  int validation_flag = _validate_ndarray(obj, shape, ndarray_shape, depth);
```
One idea to improve readability is to split this into acquiring the first array and its parameters (depth and shape), then have the next function discover all the arrays (and validate them when they're discovered), then have a third function operate on the discovered arrays.

touqir14 replied:
See https://github.com/pytorch/pytorch/pull/51731#discussion_r602715434

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+    Py_ssize_t seq_len = PySequence_Fast_GET_SIZE(seq.get());
+    if (seq_len != shape[depth]) {
+      throw ValueError("expected sequence of length %lld at dim %lld (got %lld)",
+        (long long)shape[depth], (long long)depth, (long long)seq_len);
```
Prefer explicit types (like int64_t) vs. platform-dependent types

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+
+
+template <typename T>
+std::string get_vector_str(std::vector<T> & vec, int64_t start, int64_t end=-1) {
```
Since end is never used let's simplify by leaving it out

touqir14 replied:
I think its good that ```end``` is there so that the function is general enough and might provide useful for any future use.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+}  
+
+template <typename T>
+std::string get_buffer_str(void* buffer, int64_t start, int64_t end) {
```
Let's just add one of these two printing functions. This is probably the one to keep. The vector version can call this by passing its data ptr, start and end.

touqir14 replied:
I have made ```get_vector_str``` use ```get_buffer_str``` for creating the string representation.

mruberry commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -386,6 +396,314 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  /* 
+  This utility function is used for copying elements from a raw numpy array memory to a tensor memory location by using memcpy_cpu and proper type matching.
+  */
+  return AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(at::ScalarType::Half, at::ScalarType::Bool, scalarType, "store_ndarray", [&] {
```
Nice use of the dispatch macro and OpaqueType

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,99 @@ def test_inference(default_dtype):
+        """
+
+        repeats = 10
+        np_types = list(torch_to_numpy_dtype_dict.values())
```
Use the @dtypes decorator with the corresponding torch dtypes and map them them to numpy dtypes using this dict in the test. That will instantiate a test for each dtype, which improves test granularity.

touqir14 replied:
Done.

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -1960,6 +1960,99 @@ def test_inference(default_dtype):
+        for dtype in np_types:
+            for _ in range(repeats):
+                stack = []
+                stack.append([np.array(1).astype(dtype)])
```
Instead of using repeats and calling the nested datastructure a "stack," @ngimel and I suggest handwriting the test cases in a tuple:

```
test_cases = (
  # Case 0: single c-contiguous NumPy array
  (...),
  # Case 1: single f-contiguous NumPy array
  (...),
  ...
)
```

Then each test case can be iterated over at once. The use of "repeats" seems unnecessary. It would also be helpful to add a comment for each test case. That will make it easier to understand test coverage. If a test case is especially complicated then consider using a helper function to generate it.

In addition to test cases for the functionality added in this PR, let's be extra safe and add some test cases for paths not improved by this PR to ensure they're not regressed. For example, a valid input of NumPy arrays nested as ((a, b), c). 

Once the valid inputs are working, for each error message create a "negative" test cases and use assertRaisesRegex to validate it's thrown when appropriate, too.

touqir14 replied:
Without including the optimization, ((a,b), c) throws a value error by the original slow path. I have added the negative test cases. Most of the slow path test cases are actually already there and so I added 1 negative test case for the slow path.

I think the way the test cases are laid out currently is more flexible than using a single test_cases (as shown above) without losing readability. For test cases which seem a little vague, I have added comments. I have removed ```repeats```

mruberry reviewed (commented):
Hey @touqir14!

This PR is making good progress. @ngimel and I took a look. We think the algorithm for discovering, validating, and collecting the arrays can be further simplified, and would like to see some changes in the testing, but we're increasingly happy with the approach taken in this PR. We look forward to seeing an update!

touqir14 committed:
Merge remote-tracking branch 'upstream/master'

touqir14 committed:
Simplified the code and added few more tests

touqir14 commented:
@mruberry , @ngimel , I have just pushed a new commit to address the suggestions. Its quite busy here and I am afraid I will be a little slow to respond.

mruberry commented on test/test_tensor_creation_ops.py:
```diff
@@ -2002,6 +2006,101 @@ def test_inference(default_dtype):
+
+        for arrays in tests:
+            self.assertEqual(torch.tensor(arrays, device=device), torch.from_numpy(np.stack(arrays)),
+                             exact_dtype=True, rtol=1e-05, atol=1e-08)
```
rtol=0, atol=0

mruberry commented:
Hey @touqir14! This looks pretty good; just add a skip for the failing test when necessary and take a look at extending the test cases to handle ((a, b), c).

rgommers commented:
I just closed another duplicate report for this issue, so I thought I'd check in on this. @touqir14 will you be able to update this PR?

touqir14 commented:
@rgommers , I have been a little busy in the last several weeks with my daily work and some other stuffs and my PRs have been on hold unfortunately. I will prioritize this and will try to push updates soon later in this week.

rgommers commented:
Thanks @touqir14, no worries at all, thanks for the update!

touqir14 commented:
I will be continuing with this PR. Due to my job and other projects, I was unable to work on this and I apologize for keeping this hung all this time. Can you please remove the stale tag? @mruberry @ngimel
